crates/qasm2/src/bytecode.rs

```diff
@@ -0,0 +1,352 @@
+pub struct ExprArgument {
+    #[pyo3(get)]
+    pub index: usize,
+}
+
+/// A unary operation acting on some other part of the expression tree.  This includes the `+` and
+/// `-` unary operators, but also any of the built-in scientific-calculator functions.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub struct ExprUnary {
+    #[pyo3(get)]
+    pub opcode: UnaryOpCode,
+    #[pyo3(get)]
+    pub argument: PyObject,
+}
+
+/// A binary operation acting on two other parts of the expression tree.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub struct ExprBinary {
+    #[pyo3(get)]
+    pub opcode: BinaryOpCode,
+    #[pyo3(get)]
+    pub left: PyObject,
+    #[pyo3(get)]
+    pub right: PyObject,
+}
+
+/// Some custom callable Python function that the user told us about.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub struct ExprCustom {
+    #[pyo3(get)]
+    pub callable: PyObject,
+    #[pyo3(get)]
+    pub arguments: Vec<PyObject>,
+}
+
+/// Discriminator for the different types of unary operator.  We could have a separate class for
+/// each of these, but this way involves fewer imports in Python, and also serves to split up the
+/// option tree at the top level, so we don't have to test every unary operator before testing
+/// other operations.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub enum UnaryOpCode {
+    Negate,
+    Cos,
+    Exp,
+    Ln,
+    Sin,
+    Sqrt,
+    Tan,
+}
+
+/// Discriminator for the different types of binary operator.  We could have a separate class for
+/// each of these, but this way involves fewer imports in Python, and also serves to split up the
+/// option tree at the top level, so we don't have to test every binary operator before testing
+/// other operations.
+#[pyclass(module = "qiskit._qasm2", frozen)]
+#[derive(Clone)]
+pub enum BinaryOpCode {
+    Add,
+    Subtract,
+    Multiply,
+    Divide,
+    Power,
+}
+
+/// An internal representation of the bytecode that will later be converted to the more free-form
+/// [Bytecode] Python-space objects.  This is fairly tightly coupled to Python space; the intent is
+/// just to communicate to Python as concisely as possible what it needs to do.  We want to have as
+/// little work to do in Python space as possible, since everything is slower there.
+///
+/// In various enumeration items, we use numeric keys to identify the object rather than its name.
+/// This is much more efficient in Python-space; rather than needing to build and lookup things in
+/// a hashmap, we can just build Python lists and index them directly, which also has the advantage
+/// of not needing to pass strings to Python for each gate.  It also gives us consistency with how
+/// qubits and clbits are tracked; there is no need to track both the register name and the index
+/// separately when we can use a simple single index.
+pub enum InternalBytecode {
+    Gate {
+        id: usize,
```

Comment:
Consider type aliasing `usize` to something like `object_key`.

It's taking me mental effort to actively remind myself `usize` == index to an object. Particularly when comparing e.g. `Vec<f64>` to `Vec<usize>`.

Reply:
Would you consider it more idiomatic to type alias or to newtype here?

I'm not experienced with Rust, but my gut doesn't like type aliases for renaming stuff in it because they give a vague illusion of increased type safety without actually doing anything about it - if I did
```rust
type GateId = usize;
type QubitId = usize;
let a: GateId = 0;
let b: QubitId = 2;
```
I'd still be able to type-safely write the totally nonsense statement `a + b`.

If I were going to enforce greater type safety in these types (which I could definitely be on board with), my natural inclination would be to define something like
```rust
struct GateId(usize);
struct QubitId(usize);
...
```
and impl only `Copy`, `IntoPy` and a constructor `usize -> Self` on each of them.  Is that more/less idiomatic?  It's the kind of thing I'd have done back in my F# days.

Reply:
Agreed. I think a newtype is even better than a type alias. And with Rust's zero cost abstractions, there's little downside other than a little more code.

Indeed, `struct GateId(usize);` is how you define a newtype in Rust: https://doc.rust-lang.org/rust-by-example/generics/new_types.html

But I don't think you need to define `Copy` etc, unless the compiler is erroring. For `IntoPy`, you don't need to define it on the newtype structs. Instead, in the implementation of `IntoPy<Bytecode> for InternalBytecode`, use `id.0`.

Reply:
Sorry, I was meaning derive Copy and impl the others - I was being kind of lazy, since derive is just a type of impl.

About `IntoPy` - I think it'll be marginally more efficient to impl `IntoPy` because then `Vec<QubitId>` (etc) will convert on a single iteration through them, rather than needing to `map(|x| x.0).collect()` into a covertable trait.  But otherwise, yeah for sure.

Reply:
Wrapper types added in a95e8e491.